src/server/auth/server/saml.go

```diff
@@ -0,0 +1,419 @@
+						"persisted without IDP metadata", idp.Name)
+				}
+				rawIDPMetadata = idp.SAML.IDPMetadata
+			}
+		}
+
+		// Parse IDP metadata. This code is heavily based on the
+		// crewjam/saml/samlsp.Middleware constructor
+		c.IDPMetadata = &saml.EntityDescriptor{}
+		err = xml.Unmarshal(rawIDPMetadata, c.IDPMetadata)
+		if err != nil {
+			// this comparison is ugly, but it is how the error is generated in
+			// encoding/xml
+			if err.Error() != "expected element type <EntityDescriptor> but have <EntitiesDescriptor>" {
+				return nil, fmt.Errorf("could not unmarshal EntityDescriptor from IDP metadata: %v", err)
+			}
+			// Search through <EntitiesDescriptor> & find IDP entity
+			entities := &saml.EntitiesDescriptor{}
+			if err := xml.Unmarshal(rawIDPMetadata, entities); err != nil {
+				return nil, fmt.Errorf("could not unmarshal EntitiesDescriptor from IDP metadata: %v", err)
+			}
+			for i, e := range entities.EntityDescriptors {
+				if len(e.IDPSSODescriptors) > 0 {
+					c.IDPMetadata = &entities.EntityDescriptors[i]
+					break
+				}
+			}
+			// Make sure we found an IDP entity descriptor
+			if len(c.IDPMetadata.IDPSSODescriptors) == 0 {
+				return nil, fmt.Errorf("no entity found with IDPSSODescriptor")
+			}
+		}
+	}
+
+	// Validate saml_svc_options
+	if config.SAMLServiceOptions != nil {
+		if c.IDPName == "" {
+			return nil, errors.New("cannot set saml_svc_options without configuring a SAML ID provider")
+		}
+		sso := config.SAMLServiceOptions
+		// parse ACS URL
+		if sso.ACSURL == "" {
+			return nil, errors.New("invalid SAML service options: must set ACS URL")
+		}
+		if c.ACSURL, err = url.Parse(sso.ACSURL); err != nil {
+			return nil, fmt.Errorf("could not parse SAML config ACS URL (%q): %v", sso.ACSURL, err)
+		} else if c.ACSURL.Scheme == "" {
+			return nil, fmt.Errorf("ACS URL %q is invalid (no scheme)", sso.ACSURL)
+		}
+
+		// parse Metadata URL
+		if sso.MetadataURL == "" {
+			return nil, errors.New("invalid SAML service options: must set Metadata URL")
+		}
+		if c.MetadataURL, err = url.Parse(sso.MetadataURL); err != nil {
+			return nil, fmt.Errorf("could not parse SAML config metadata URL (%q): %v", sso.MetadataURL, err)
+		} else if c.MetadataURL.Scheme == "" {
+			return nil, fmt.Errorf("Metadata URL %q is invalid (no scheme)", sso.MetadataURL)
+		}
+
+		// parse Dash URL
+		if sso.DashURL != "" {
+			if c.DashURL, err = url.Parse(sso.DashURL); err != nil {
+				return nil, fmt.Errorf("could not parse Pachyderm dashboard URL (%q): %v", sso.DashURL, err)
+			} else if c.DashURL.Scheme == "" {
+				return nil, fmt.Errorf("Pachyderm dashboard URL %q is invalid (no scheme)", sso.DashURL)
+			}
+		}
+	}
+
+	return c, nil
+}
+
+// tryUpdateConfig validates 'config', and if it valides successfully, loads it
+// into the apiServer's config cache. The caller should already hold a.configMu
+// and a.samlSPMu (as this updates a.samlSP)
+func (a *apiServer) tryUpdateConfig(config *auth.AuthConfig) error {
```

Comment:
Could we just call this `updateConfig` it seems like it's trying in the sense that it might fail and return an error. But that's normal enough behavior that I don't think you need to specify it in the function name.

Reply:
Done